File: eval-gen/src/fact-extractor.ts
import { Fact, DatasetProfile, ColumnProfile } from './types';

export interface ExtractFactsOptions {
  /** Hard cap on total facts emitted (default 200) */
  maxFacts?: number;
  /** Target distinct records to sample (default scales with maxFacts; floor 50) */
  targetRecords?: number;
  /** Cap facts emitted per record (default unlimited) */
  maxFactsPerRecord?: number;
}

/**
 * Extract atomic facts from dataset records using stratified sampling.
 * Selects diverse records covering common values, rare values, extremes, and nulls.
 *
 * Accepts either a number (legacy `maxFacts`) or an options object so that
 * `targetRecords` can scale independently — wide-schema datasets otherwise
 * exhaust `maxFacts` after only a handful of rows.
 */
export function extractFacts(
  records: Record<string, unknown>[],
  profile: DatasetProfile,
  optionsOrMaxFacts: number | ExtractFactsOptions = 200,
): Fact[] {
  const opts: ExtractFactsOptions =
    typeof optionsOrMaxFacts === 'number' ? { maxFacts: optionsOrMaxFacts } : optionsOrMaxFacts;
  const maxFacts = opts.maxFacts ?? 200;
  const explicitTargetRecords = opts.targetRecords;
  const maxFactsPerRecord = opts.maxFactsPerRecord ?? Number.POSITIVE_INFINITY;

  const facts: Fact[] = [];
  const selectedIndices = selectStratifiedIndices(
    records,
    profile,
    maxFacts,
    explicitTargetRecords,
  );

  let factId = 0;
  for (const rowIndex of selectedIndices) {
    const record = records[rowIndex];
    const fileLabel = record._source_file ? String(record._source_file) : profile.fileName;
    const rowRef = `${fileLabel}:row ${rowIndex + 1}`;

    let factsForThisRecord = 0;
    for (const col of profile.columns) {
      if (factsForThisRecord >= maxFactsPerRecord) break;
      const value = record[col.name];
      if (value === null || value === undefined || value === '') continue;
      if (col.name === '_source_file') continue;

      facts.push({
        id: `f-${++factId}`,
        field: col.name,
        value,
        rowReference: rowRef,
        record: { ...record },
      });
      factsForThisRecord++;
    }

    if (facts.length >= maxFacts) break;
  }

  return facts.slice(0, maxFacts);
}

/**
 * Select record indices using stratified sampling:
 * - Common values (most frequent category values)
 * - Rare values (least frequent category values)
 * - Extreme values (min/max of numeric columns)
 * - Null-heavy records (records with missing fields)
 * - Evenly spaced for general coverage
 */
function selectStratifiedIndices(
  records: Record<string, unknown>[],
  profile: DatasetProfile,
  maxFacts: number,
  explicitTargetRecords?: number,
): number[] {
  // Decouple row count from column width so wide schemas don't starve the row pool.
  // Default floor of 50 records when available; callers can pass an explicit target.
  const defaultTarget = Math.max(50, Math.ceil(maxFacts / 5));
  const targetRecords = Math.min(
    records.length,
    explicitTargetRecords ?? defaultTarget,
  );
  const selected = new Set<number>();

  // 1. Records with extreme numeric values
  for (const col of profile.columns) {
    if (col.dataType === 'number' && col.min !== undefined && col.max !== undefined) {
      const minIdx = records.findIndex(r => Number(r[col.name]) === col.min);
      const maxIdx = records.findIndex(r => Number(r[col.name]) === col.max);
      if (minIdx >= 0) selected.add(minIdx);
      if (maxIdx >= 0) selected.add(maxIdx);
    }
  }

  // 2. Records from rare categories
  for (const col of profile.columns) {
    if (col.valueCounts) {
      const sorted = Object.entries(col.valueCounts).sort((a, b) => a[1] - b[1]);
      // Rarest category
      if (sorted.length > 0) {
        const rarest = sorted[0][0];
        const idx = records.findIndex(r => String(r[col.name]) === rarest);
        if (idx >= 0) selected.add(idx);
      }
      // Most common category
      if (sorted.length > 1) {
        const most = sorted[sorted.length - 1][0];
        const idx = records.findIndex(r => String(r[col.name]) === most);
        if (idx >= 0) selected.add(idx);
      }
    }
  }

  // 3. Records with null/empty fields
  const nullCols = profile.columns.filter(c => c.nullCount > 0);
  for (const col of nullCols.slice(0, 3)) {
    const idx = records.findIndex(r =>
      r[col.name] === null || r[col.name] === undefined || r[col.name] === ''
    );
    if (idx >= 0) selected.add(idx);
  }

  // 4. Evenly spaced fill
  const remaining = targetRecords - selected.size;
  if (remaining > 0) {
    const step = Math.max(1, Math.floor(records.length / remaining));
    for (let i = 0; i < records.length && selected.size < targetRecords; i += step) {
      selected.add(i);
    }
  }

  return Array.from(selected).sort((a, b) => a - b);
}

/**
 * Group facts by row reference for easier question generation
 */
export function groupFactsByRecord(facts: Fact[]): Map<string, Fact[]> {
  const groups = new Map<string, Fact[]>();
  for (const fact of facts) {
    const existing = groups.get(fact.rowReference) ?? [];
    existing.push(fact);
    groups.set(fact.rowReference, existing);
  }
  return groups;
}

/**
 * Get a summary of facts for LLM context (limits token usage).
 * Each fact line is prefixed with its stable [f-N] ID so the LLM can cite specific facts.
 */
export function summarizeFacts(facts: Fact[], maxRecords: number = 15): string {
  const grouped = groupFactsByRecord(facts);
  const lines: string[] = [];

  let count = 0;
  for (const [rowRef, recordFacts] of grouped) {
    if (count >= maxRecords) break;

    const fields = recordFacts
      .map(f => `[${f.id}] ${f.field}=${JSON.stringify(f.value)}`)
      .join(', ');
    lines.push(`[${rowRef}] ${fields}`);
    count++;
  }

  return lines.join('\n');
}

File: eval-ui-winui3/src/EvalToolkit.EvalGen/Pipeline/FactExtractor.cs
using System.Globalization;
using System.Text.Json;
using EvalToolkit.Core;

namespace EvalToolkit.EvalGen.Pipeline;

/// <summary>
/// Port of <c>eval-gen/src/fact-extractor.ts</c>. Extracts atomic facts from
/// dataset records using stratified sampling; provides helpers to group and
/// summarize facts for LLM context.
/// </summary>
public static class FactExtractor
{
    /// <summary>Options for <see cref="ExtractFacts"/>.</summary>
    public sealed record ExtractFactsOptions
    {
        /// <summary>Hard cap on total facts emitted (default 200).</summary>
        public int MaxFacts { get; init; } = 200;

        /// <summary>Target distinct records to sample (default scales with MaxFacts; floor 50).</summary>
        public int? TargetRecords { get; init; }

        /// <summary>Cap facts emitted per record (default unlimited).</summary>
        public int MaxFactsPerRecord { get; init; } = int.MaxValue;
    }

    /// <summary>
    /// Extract atomic facts from dataset records using stratified sampling.
    /// Mirrors TS <c>extractFacts</c>.
    /// </summary>
    public static IReadOnlyList<Fact> ExtractFacts(
        IReadOnlyList<IReadOnlyDictionary<string, object?>> records,
        DatasetProfile profile,
        ExtractFactsOptions? options = null)
    {
        options ??= new ExtractFactsOptions();
        var maxFacts = options.MaxFacts;
        var maxFactsPerRecord = options.MaxFactsPerRecord;

        var facts = new List<Fact>();
        var selectedIndices = SelectStratifiedIndices(records, profile, maxFacts, options.TargetRecords);

        int factId = 0;
        foreach (var rowIndex in selectedIndices)
        {
            var record = records[rowIndex];
            string fileLabel = profile.FileName;
            if (record.TryGetValue("_source_file", out var sf) && sf is not null)
            {
                // Parity with TS truthiness: empty string falls back to profile.FileName.
                var sourceFileLabel = Profiler.ConvertToString(sf);
                if (!string.IsNullOrEmpty(sourceFileLabel))
                {
                    fileLabel = sourceFileLabel;
                }
            }
            var rowRef = $"{fileLabel}:row {rowIndex + 1}";

            int factsForThisRecord = 0;
            foreach (var col in profile.Columns)
            {
                if (factsForThisRecord >= maxFactsPerRecord) break;
                record.TryGetValue(col.Name, out var value);
                if (value is null || (value is string s && s.Length == 0)) continue;
                if (col.Name == "_source_file") continue;

                facts.Add(new Fact
                {
                    Id = $"f-{++factId}",
                    Field = col.Name,
                    Value = value,
                    RowReference = rowRef,
                    Record = CloneRecord(record),
                });
                factsForThisRecord++;
            }

            if (facts.Count >= maxFacts) break;
        }

        return facts.Count <= maxFacts ? facts : facts.Take(maxFacts).ToList();
    }

    private static Dictionary<string, object?> CloneRecord(IReadOnlyDictionary<string, object?> record)
    {
        var copy = new Dictionary<string, object?>(record.Count, StringComparer.Ordinal);
        foreach (var kv in record) copy[kv.Key] = kv.Value;
        return copy;
    }

    private static List<int> SelectStratifiedIndices(
        IReadOnlyList<IReadOnlyDictionary<string, object?>> records,
        DatasetProfile profile,
        int maxFacts,
        int? explicitTargetRecords)
    {
        var defaultTarget = Math.Max(50, (int)Math.Ceiling(maxFacts / 5.0));
        var targetRecords = Math.Min(records.Count, explicitTargetRecords ?? defaultTarget);
        var selected = new SortedSet<int>();

        // 1. Records with extreme numeric values
        foreach (var col in profile.Columns)
        {
            if (col.DataType == "number" && col.Min is not null && col.Max is not null)
            {
                var minTarget = Convert.ToDouble(col.Min, CultureInfo.InvariantCulture);
                var maxTarget = Convert.ToDouble(col.Max, CultureInfo.InvariantCulture);
                int minIdx = FindIndex(records, r =>
                {
                    r.TryGetValue(col.Name, out var v);
                    return v is not null && Profiler.TryParseNumber(Profiler.ConvertToString(v), out var n) && n == minTarget;
                });
                int maxIdx = FindIndex(records, r =>
                {
                    r.TryGetValue(col.Name, out var v);
                    return v is not null && Profiler.TryParseNumber(Profiler.ConvertToString(v), out var n) && n == maxTarget;
                });
                if (minIdx >= 0) selected.Add(minIdx);
                if (maxIdx >= 0) selected.Add(maxIdx);
            }
        }

        // 2. Records from rare/most common categories
        foreach (var col in profile.Columns)
        {
            if (col.ValueCounts is not null)
            {
                var sorted = col.ValueCounts.OrderBy(kv => kv.Value).ToList();
                if (sorted.Count > 0)
                {
                    var rarest = sorted[0].Key;
                    int idx = FindIndex(records, r =>
                    {
                        r.TryGetValue(col.Name, out var v);
                        return string.Equals(Profiler.ConvertToString(v), rarest, StringComparison.Ordinal);
                    });
                    if (idx >= 0) selected.Add(idx);
                }
                if (sorted.Count > 1)
                {
                    var most = sorted[^1].Key;
                    int idx = FindIndex(records, r =>
                    {
                        r.TryGetValue(col.Name, out var v);
                        return string.Equals(Profiler.ConvertToString(v), most, StringComparison.Ordinal);
                    });
                    if (idx >= 0) selected.Add(idx);
                }
            }
        }

        // 3. Records with null/empty fields (up to 3 columns)
        var nullCols = profile.Columns.Where(c => c.NullCount > 0).Take(3).ToList();
        foreach (var col in nullCols)
        {
            int idx = FindIndex(records, r =>
            {
                r.TryGetValue(col.Name, out var v);
                return v is null || (v is string s && s.Length == 0);
            });
            if (idx >= 0) selected.Add(idx);
        }

        // 4. Evenly spaced fill
        var remaining = targetRecords - selected.Count;
        if (remaining > 0)
        {
            var step = Math.Max(1, records.Count / remaining);
            for (int i = 0; i < records.Count && selected.Count < targetRecords; i += step)
            {
                selected.Add(i);
            }
        }

        return selected.ToList();
    }

    private static int FindIndex<T>(IReadOnlyList<T> list, Func<T, bool> predicate)
    {
        for (int i = 0; i < list.Count; i++)
        {
            if (predicate(list[i])) return i;
        }
        return -1;
    }

    /// <summary>
    /// Group facts by row reference for easier question generation. Preserves
    /// insertion order via <c>Dictionary&lt;,&gt;</c> (.NET dictionaries
    /// preserve insertion order, matching JS Map semantics).
    /// </summary>
    public static IReadOnlyDictionary<string, IReadOnlyList<Fact>> GroupFactsByRecord(IEnumerable<Fact> facts)
    {
        var groups = new Dictionary<string, List<Fact>>(StringComparer.Ordinal);
        foreach (var fact in facts)
        {
            if (!groups.TryGetValue(fact.RowReference, out var list))
            {
                list = new List<Fact>();
                groups[fact.RowReference] = list;
            }
            list.Add(fact);
        }

        var result = new Dictionary<string, IReadOnlyList<Fact>>(groups.Count, StringComparer.Ordinal);
        foreach (var kv in groups) result[kv.Key] = kv.Value;
        return result;
    }

    /// <summary>
    /// Get a summary of facts for LLM context. Each fact line is prefixed with
    /// its stable [f-N] ID so the LLM can cite specific facts. Mirrors TS
    /// <c>summarizeFacts</c>.
    /// </summary>
    public static string SummarizeFacts(IEnumerable<Fact> facts, int maxRecords = 15)
    {
        var grouped = GroupFactsByRecord(facts);
        var lines = new List<string>();
        int count = 0;
        foreach (var (rowRef, recordFacts) in grouped)
        {
            if (count >= maxRecords) break;
            var fields = string.Join(", ", recordFacts.Select(f =>
                $"[{f.Id}] {f.Field}={JsonStringify(f.Value)}"));
            lines.Add($"[{rowRef}] {fields}");
            count++;
        }
        return string.Join("\n", lines);
    }

    /// <summary>
    /// JS <c>JSON.stringify</c>-equivalent for primitives. Strings get double
    /// quotes with backslash escapes; numbers/bools/null get their wire form;
    /// objects/arrays get System.Text.Json serialization.
    /// </summary>
    internal static string JsonStringify(object? value)
    {
        if (value is null) return "null";
        if (value is bool b) return b ? "true" : "false";
        if (value is string s) return JsonSerializer.Serialize(s);
        if (value is double d)
        {
            if (double.IsNaN(d) || double.IsInfinity(d)) return "null";
            return Profiler.ConvertToString(d);
        }
        if (value is float f)
        {
            if (float.IsNaN(f) || float.IsInfinity(f)) return "null";
            return Profiler.ConvertToString(f);
        }
        if (value is decimal m) return m.ToString(CultureInfo.InvariantCulture);
        if (value is sbyte or byte or short or ushort or int or uint or long or ulong)
        {
            return Convert.ToInt64(value, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture);
        }
        return JsonSerializer.Serialize(value);
    }
}
